### PR TITLE
Manage maven cache size

### PR DESCRIPTION
**Pull request by ericbottard**
Introduce a wrapper around an existing ResourceLoader that limits the
total size used on disk, purging LRU entries first.

Fixes https://github.com/spring-cloud/spring-cloud-deployer/issues/131

### CODE REVIEW

File: spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/EOFInputStream.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package org.springframework.cloud.deployer.resource.support;
```
Comment (ilayaperumalg) [inline spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/EOFInputStream.java]: Missing copyright.

File: spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/EOFInputStream.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Created by ericbottard on 08/12/16.
```
Comment (ilayaperumalg) [inline spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/EOFInputStream.java]: Would be a good idea to change this in your IDE to have `@author` etc.,

File: spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/LRUResourceLoader.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
+package org.springframework.cloud.deployer.resource.support;
```
Comment (ilayaperumalg) [inline spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/LRUResourceLoader.java]: copyright.

**Comment by ericbottard** [comment]
Added copyright where appropriate

**Comment by jvalkeal** [comment]
I don't think this will work in a real life:

1. `Resource` may not have a backing physical `File`
2. With maven repo you should purge all files related to an artifact(pom,sha, etc).
3. Having multiple caches in a same system will start to fight with each others, right?

**Comment by ericbottard** [comment]
1. This is meant to wrap our DelegatingResourceLoader, so each resource *will* be a File
2. True in theory, but a) this may apply to non-maven resources. b) we only want to avoid crash in resource constraint envs, not be 100% perfect maven-wise. There may be a problem with timestamped SNAPSHOTs though.
3. I don't get what you're saying

**Comment by ericbottard** [comment]
I know this is not an ideal solution, serves more as a basis for discussion (which already started with @markpollack )

**Comment by jvalkeal** [comment]
Yeah, this is a bit complex issue. I don't want any external system to remove stuff from my local m2 repo, that's way too invasive. Back in days there used to be a lot of issues with gradle if jar artifact was removed but metadata files told file is there. What about if that file in currently in use by some other process?

Maybe this should not be a public impl?

**Comment by ericbottard** [comment]
This is mainly for use in CF, not for a "laptop" m2 cache

**Comment by jvalkeal** [comment]
ah, ok. I surely don't care about m2 on CF(until someone starts to use that dave's thin wrapper).

File: spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/LRUResourceLoader.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+			super(5, 0.75f, true);
+			this.maxSize = maxSize;
+		}
+
+
+		@Override
+		protected boolean removeEldestEntry(Map.Entry<File, Long> eldest) {
+			// Compute cumulative used disk space, in MRU order
+			List<Map.Entry<File, Long>> lruLast = new ArrayList<>(this.entrySet());
+			Collections.reverse(lruLast);
```
Comment (markpollack) [inline spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/LRUResourceLoader.java]: why reverse the entries as they are stored by insertion order?

Comment (ericbottard) [inline spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/LRUResourceLoader.java]: They're actually stored in last access order (LRU first), see 3rd parameter to the constructor (and associated javadoc). So I want them in reverse order to have the bigger sums represent the LRU items

**Comment by markpollack** [comment]
1. We can inforce the impl to only deal with files to make it explicit.  We can also just put this as a protected class bandaid in the cf-deployer code base to limit the scope of usage since it is only for CF. ATM
2. It would seem deleting everything under the .m2 version directory to remove all maven artifacts would be the cleaner approach.
3. " I don't want any external system to remove stuff from my local m2 repo", well you do care if it means the SCDF Server on PCF can only run for a couple weeks before running out of disk space.

File: spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/LRUResourceLoader.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	private final ClassLoader classLoader = ClassUtils.getDefaultClassLoader();
+
+	private final ResourceLoader delegate;
+
+	private final Map<File, Long> cache;
+
+	/**
+	 * Wrap an existing ResourceLoader, limiting total used space to {@literal maxSize} bytes.
+	 */
+	public LRUResourceLoader(ResourceLoader delegate, long maxSize) {
```
Comment (markpollack) [inline spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/LRUResourceLoader.java]: I don't see how we can know the size of this specific file up front to set it to an appropriate value.  Consider the case where we have 1GB of free disk.  I should be able to have a 500MB .jar file and 10 * 50MB other artfiacts or have 100 * 50 MB file.  How to set the maxSize in this case?

Comment (ericbottard) [inline spring-cloud-deployer-resource-support/src/main/java/org/springframework/cloud/deployer/resource/support/LRUResourceLoader.java]: maxSize here is the total combined cache size. So set it to 1GB in your case

**Comment by markpollack** [comment]
Here is an alternative implementation after a google hangout with Eric that should help solve the issues of snapshot and also proper cleanup of directories (vs. just files)

1. We use the config property ${maven.localRepository} to determine the path where the maven cache is located.  This will later be used to determine the free disk space on the underlying partition for that path.
2. The LRU cache will only keep track of filenames that were most recently used, no bookkeeping of file size.  This is due to poor bookkeeping when mvn downloads snapshots - there is an extra copy of the latest file even when just one snapshot is downloaded and the bookkeeping increases inaccuracy when there are older snapshots laying around.
3. As a ctor argument we specify the low water mark of free disk space that will trigger the cleanup operation.  This will be expressed as a percentage, default can be 10%.  
4. After a resource is downloaded, we check if the low water mark of free disk space has been exceeded, iterate through the LRU cache and delete the parent directory of the LRU files until the low water mark is not exceeded.
5. Move this this to spring-cloud-dataflow-server-cloudfoundry repo and provide a new instance of DelegatingResourceLoader to override what is provided by `DataFlowControllerAutoConfiguration` (there is a `@ConditionalOnBean` annotation in the autoconfig of the dataflow core.

**Comment by ericbottard** [comment]
Closing, will open a PR against https://github.com/spring-cloud/spring-cloud-dataflow-server-cloudfoundry

**Pull request closed without merging**